### PR TITLE
[ticket/16543] Update query-string in stylesheet.css to hashes

### PR DESCRIPTION
**Pull request by 3D-I**
PHPBB3-16543

Checklist:

- [x] Correct branch: master for new features; 3.3.x & 3.2.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html), [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html) and [3.2.x](https://area51.phpbb.com/docs/dev/3.2.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-16543

### CODE REVIEW

**Comment by 3D-I** [comment]
Running the script will replace querystrings with hashes


immagine

**Comment by DavidIQ** [comment]
I don't think the `patch_` prefix goes with the file.

**Comment by 3D-I** [comment]
> 
> 
> I don't think the `patch_` prefix goes with the file name.

Good catch, thx. :+1:

**Comment by hanakin** [comment]
As we will not be using imports in 4.0 this should only be required in 3.x. It should have no conflicts either as the name of the loaded file will be theme.css just for clarification. So with that said pending test that it’s working looks/sounds good to me.

**Comment by 3D-I** [comment]
There are no tests for the `build/` utilities/tools afaik.

**Comment by rxu** [comment]
What if we use `INCLUDECSS` instead?

**Comment by 3D-I** [comment]
I'd need to read an example first.

**Comment by DavidIQ** [comment]
The way `T_STYLESHEET_LINK` gets generated is already kind of what `INCLUDECSS` does so it won't really help.

**Comment by rxu** [comment]
> Where should it be the difference since the issue is in the browser of the board members?
It will add `assets_version` query.

> The way T_STYLESHEET_LINK gets generated is already kind of what INCLUDECSS does so it won't really help.
It's only applied to the `stylesheet.css`. What I'm talking about is adding every single css file with includecss (f.e. into `stylesheet.html` and then include this into `overall_header.html`).

**Comment by 3D-I** [comment]
Well, I see now what you mean. 
Whatever you guys prefer we can do as long as the mission is accomplished.

**Comment by DavidIQ** [comment]
> It's only applied to the `stylesheet.css`. What I'm talking about is adding every single css file with includecss (f.e. into `stylesheet.html` and then include this into `overall_header.html`).

I figured that's what you were suggesting, but wasn't sure because we wouldn't do such a change in a current release. I think it's too invasive.

**Comment by rxu** [comment]
No problem. The point is just to have an alternative way.

**Comment by hanakin** [comment]
It’s also horrible for performance

**Comment by 3D-I** [comment]
I have it already manually tested, see screenshots.

**Comment by JoshyPHP** [comment]
> I don't think the `patch_` prefix goes with the file name.

It was part of it, but you can name it anything you like, I haven't looked up whether phpBB has a convention for those files or where they're supposed to reside. I generally recommend starting with a verb that describes what the file does. `patch_` or `update_` are good candidates.

**Comment by marc1706** [comment]
So if I understand this correctly, the script will always create a query string with `?hash=abcdefg` while you currently updated this to have `?v=3.3.1`? I would actually go with updating the query string for every release, i.e. run this as part of the release process. So for that we could have this even easier and just have the script replace for example `?v=3.3` with the version passed to the script which would be 3.3.1 here.

**Comment by hanakin** [comment]
Or do we even need the version stuff? Just append random hash to end of every file on build that’s based on the release #

**Comment by JoshyPHP** [comment]
If you use a hash, it means the script can be run any number of times with no side effects. It can be run during every build including development builds.

Using a version number means the cache gets nuked even when there's no change and on the other hand it's not invalidated if the file changes but the version doesn't, for whatever reason. It's just more efficient and more reliable to use a hash.

**Comment by marc1706** [comment]
Then let's do it with a hash right away

**Comment by 3D-I** [comment]
> you currently updated this to have `?v=3.3.1`? I

The first proposal was to update the version number (see discussion on the ticket) then @JoshyPHP provided a script, I didn't use hashes because obviously the version number would have been changed during the build anyway (see running the script).

File: build/update_stylesheet_querystrings.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	array_map('patch_file', glob($glob));
+}
+
+function patch_file(string $filepath): void
+{
+	$file	= file_get_contents($filepath);
+	$old	= $file;
+	$new	= preg_replace_callback(
+		'(^@import\\s+url\\([\'"](?<basename>\\w++\\.css)\\?\\K(?:hash|v)=[^\'"]++)m',
+		function ($m) use ($filepath)
```
Comment (marc1706) [inline build/update_stylesheet_querystrings.php:40]:
```suggestion
		function ($match) use ($filepath)
```

**Pull request merged**